tests/spell_check3.k

```diff
@@ -0,0 +1,318 @@
+    role SpellChecker {
+        public void advanceToNextWord() {
+            if(isFinished()) return;
+
+            // Advance past current word
+            while(isAlphabetic(currentCharacter())) {
+                setPosition(currentPosition() + 1);
+                if(isFinished()) return;
+            }
+
+            // Advance past non-alphabetic characters
+            while(!isAlphabetic(currentCharacter())) {
+                setPosition(currentPosition() + 1);
+                if(isFinished()) return;
+            }
+
+            Dictionary.checkCurrentWord();
+        }
+
+        public String currentWord() {
+            if(isFinished()) return "";
+
+            String output = "";
+            int pos = currentPosition().clone;
+
+            while(isAlphabetic(Text.characterAt(pos))) {
+                output = output + Text.characterAt(pos++);
+                // NOTE: Cannot call Text.getText().length() here.
+                if(pos >= Text.length()) return output;
+            }
+
+            return output;
+        }
+
+        private String currentCharacter() {
+            return Text.characterAt(currentPosition());
+        }
+
+        private boolean isFinished() {
+            return currentPosition() >= Text.length();
+        }
+
+        private boolean isAlphabetic(String character) {
+            assert(character.length() == 1, "character string length not equal to one.");
+            return character == "-" || 
+                   (character >= "a" && character <= "z") ||
+                   (character >= "A" && character <= "Z");
+        }
+    } requires {
+        int currentPosition() const;
+        void setPosition(int pos);
+    }
+
+    /////////////////////////////////////////
+    role Text {
+        public String characterAt(int pos) {
+            return getText().substring(pos, pos+1);
+        }
+        public int length() {
+            return getText().length();
+        }
+    } requires {
+        String getText() const;
+        void splice(int startIndex, int endIndex, String text);
+    }
+
+    /////////////////////////////////////////
+    stageprop Dictionary {
+        public void checkCurrentWord() {
+            String word = SpellChecker.currentWord();
+            List<String> suggestions = suggestSpelling(word);
+
+            if(suggestions.size() > 0) {
+                int pos = SpellChecker.currentPosition();
+
+                // TODO: User input and selection
+                Text.splice(pos, pos + word.length(), "[" + suggestions.get(0) + "]");
+            }
+
+            SpellChecker.advanceToNextWord();
+        }
+    } requires {
+        List<String> suggestSpelling(String word) const;
+    }
+
+    /////////////////////////////////////////
+
+    public SpellCheck(Text text) {
+        Text = text;
+        SpellChecker = this;
+        Dictionary = new SimpleDictionary();
+    }
+
+    public void start() {
+        setPosition(0);
```

Comment:
I'm a bit confused, the reason I'm trying hard to stay away from any programmer concepts like iterators and cursors, is that they aren't specified in the use case. Even the text role isn't there, but I did consider making it the SpellChecker instead. But then the text buffer, the selection, the text file or any other object playing that role needs to have an iterator for spell checking?

I'm trying to stay faithful to the use case, but I don't know what I'm missing here.